server/asset/eth/coiner.go

```diff
@@ -0,0 +1,281 @@
+//
+// In the case of ethereum it is extra important to check confirmations before
+// confirming a swap. Even though we check the initial transaction's data, if
+// that transaction were in mempool at the time, it could be swapped out with
+// any other values if a user sent another transaction with a higher gas fee
+// and the same account and nonce, effectivly voiding the transaction we
+// expected to be mined.
+func (c *swapCoin) Confirmations(_ context.Context) (int64, error) {
+	swap, err := c.backend.node.swap(c.backend.rpcCtx, c.secretHash)
+	if err != nil {
+		return -1, err
+	}
+
+	assertMempool := func() error {
+		_, mempool, err := c.backend.node.transaction(c.backend.rpcCtx, common.HexToHash(c.txid))
+		if err != nil {
+			return fmt.Errorf("unable to fetch transaction: %v", err)
+		}
+		if !mempool {
+			return fmt.Errorf("critical problem with swap coin: tx not in mempool but swap state is %s", SwapState(swap.State))
+		}
+		return nil
+	}
+
+	switch c.sct {
+	case sctRedeem:
+		// There should be no need to check the counter party, or value
+		// as a swap with a specific secret hash that has been redeemed
+		// wouldn't have been redeemed without ensuring the initiator
+		// is the expected address and value was also as expected. Also
+		// not validating the locktime, as the swap is redeemed and
+		// locktime no longer relevant.
+		ss := SwapState(swap.State)
+		if ss == SSRedeemed {
+			// While not completely accurate, we know that if the
+			// swap is redeemed the redemption has at least one
+			// confirmation.
+			return 1, nil
+		}
+		// If swap is in the Initiated state, the transaction must be
+		// unmined.
+		if ss == SSInitiated {
+			if err := assertMempool(); err != nil {
+				return -1, err
+			}
+			// Assume the tx still has a chance of being mined.
+			return 0, nil
+		}
+		// If swap is in None state, then the redemption can't possibly
+		// succeed as the swap must already be in the Initialized state
+		// to redeem. If the swap is in the Refunded state, then the
+		// redemption either failed or never happened.
+		return -1, fmt.Errorf("redemption in failed state with swap at %s state", ss)
+
+	case sctInit:
+		// Uninitiated state is zero confs. It could still be in mempool.
+		// It is important to only trust confirmations according to the
+		// swap contract. Until there are confirmations in the contract,
+		// we cannot be sure that initiation happened successfuly. If
+		// the swap has zero confirmations but the init transaction has
+		// confirmations, something is wrong or initialization failed
+		// and this contract should be dropped.
+		if SwapState(swap.State) == SSNone {
+			if err := assertMempool(); err != nil {
+				return -1, err
+			}
+			// Assume the tx still has a chance of being mined.
+			return 0, nil
+		}
+		// Any other swap state is ok. We are sure that initialization
+		// happened.
+
+		// The contract has some number of confirmations, and we are
```

Comment:
contract -> swap

Reply:
Mmm, yeah, the "contract" language was too utxo specific here.  "swap" as you suggest is probably best, but I wouldn't object to "swap initiation transaction" in places that you don't mind being more explicit, but probably that would be needlessly wordy to say everywhere.

Reply:
Ahh, I don't know. Transactions still gotta happen... We are requiring the transaction tx, though we could not, it's probably good to get things started in mempool like other coins. Also we don't save the "txfee" in the contract, we could, but having the initial tx works too.

Reply:
Changed here.